src/main/java/by/triumgroup/recourse/service/impl/CourseServiceImpl.java

```diff
@@ -1,49 +1,95 @@
+import java.util.*;
 
 import static by.triumgroup.recourse.util.RepositoryCallWrapper.wrapJPACall;
+import static by.triumgroup.recourse.util.RepositoryCallWrapper.wrapJPACallToOptional;
 
+@SuppressWarnings("OptionalGetWithoutIsPresent")
 public class CourseServiceImpl
         extends AbstractCrudService<Course, Integer>
         implements CourseService {
 
-    private final CourseRepository repository;
+    private final CourseRepository courseRepository;
+    private final UserRepository userRepository;
 
-    public CourseServiceImpl(CourseRepository repository) {
-        super(repository);
-        this.repository = repository;
+    public CourseServiceImpl(CourseRepository courseRepository, UserRepository userRepository) {
+        super(courseRepository);
+        this.courseRepository = courseRepository;
+        this.userRepository = userRepository;
     }
 
     @Override
     public List<Course> searchByTitle(String title, Pageable pageable) {
-        return wrapJPACall(() -> repository.findByTitleContainingIgnoreCaseOrderByIdDesc(title, pageable));
+        return wrapJPACall(() -> courseRepository.findByTitleContainingIgnoreCaseOrderByIdDesc(title, pageable));
     }
 
     @Override
     public List<Course> findByStatus(Course.Status status, Pageable pageable) {
-        return wrapJPACall(() -> repository.findByStatusOrderByIdDesc(status, pageable));
+        return wrapJPACall(() -> courseRepository.findByStatusOrderByIdDesc(status, pageable));
     }
 
     @Override
-    public boolean registerStudentToCourse(Integer studentId, Course courseId) {
-//        Set<User> registeredStudents = course.getStudents();
-//        registeredStudents.add(student);
-//        wrapJPACall(() -> repository.save(course))
-        throw new NotImplementedException();
+    public void registerStudentToCourse(Integer studentId, Integer courseId) {
+        Optional<Course> courseOptional = wrapJPACallToOptional(() -> courseRepository.findOne(courseId));
+        Optional<User> user = wrapJPACallToOptional(() -> userRepository.findOne(studentId));
+        List<ErrorMessage> errorMessages = new ArrayList<>();
+        if (!courseOptional.isPresent()) {
+            errorMessages.add(new ErrorMessage("course", "Not found"));
+        }
+        if (!user.isPresent()) {
+            errorMessages.add(new ErrorMessage("user", "Not found"));
+        }
+        if (!errorMessages.isEmpty()) {
+            throw new ServiceNotFoundException(errorMessages);
+        }
+        if (user.get().getRole() != User.Role.STUDENT) {
+            errorMessages.add(new ErrorMessage("user", "User must be a student"));
+        }
+        Course course = courseOptional.get();
+        if (course.getStatus() != Course.Status.REGISTRATION) {
+            errorMessages.add(new ErrorMessage("course", "Course must has registration status"));
+        }
+        if (!errorMessages.isEmpty()) {
+            throw new ServiceBadRequestException(errorMessages);
```

Comment:
Consider create a new method for this exception throwing to hide, that we use exceptions to handle basic business-logic issues...